### DISCUSSION TITLE
Adding a new variant for autocompletion

### DISCUSSION THREAD

**Discussion by wojciech-kulik** [L3MON4D3/LuaSnip#1058] [Q&A] [answered]
Hello,

I'm working with Swift language, I would like to achieve some transformation for closures.

Let's say I've got the function:

```swift
func foo(value: String, completion: () -> ()) {
    print("test")
}
```

When I use autocompletion I get:
<img width="464" alt="image" src="https://github.com/L3MON4D3/LuaSnip/assets/3128467/0c547b0c-7c46-4d55-9a24-cc165972cb9f">

If I hit enter when closure (`() -> ()`) is selected I would like it to be transformed into:
<img width="312" alt="image" src="https://github.com/L3MON4D3/LuaSnip/assets/3128467/1cb5eb1e-b70f-4a99-abf9-d141260acad5">

any guidance on how I could achieve that would be great.

Should I do it manually in lua overriding `<cr>` hit in selection mode? Or are there any mechanisms from LuaSnip/nvim-cmp which could make it simpler?

Thanks in advance!

**Comment by L3MON4D3** [comment]
Hi!

The regular approach to changing the snippet manually like that would be a choiceNode, which means you couldn't use `<Cr>`, but some other keycombo.

I guess the other issue is intercepting the completion-result from nvim-cmp/lsp.. do you know where the snippet for the function is coming from?

**Reply by wojciech-kulik** [reply]
I guess it is coming from LSP, because it's based on the function that I declared, but I'm not sure how it works internally.

You mentioned choiceNode, but if it's provided by LSP, can I add somehow a choiceNode?

It sounds like it would be easier to just run transformation on buffer line, regex to find the pattern and gsub to replace it :D

**Reply by L3MON4D3** [reply, marked as answer]
Ah, okay, then this is a bit tricky:
You'd have to intercept the results from lsp and modify them, [I do something like that](https://github.com/L3MON4D3/Dotfiles/blob/master/.config/nvim/lua/plugins/lspconfig.lua#L68-L98) to solve a pretty similar problem (for c++, add a choiceNode to toggle between `()` and `{}`).

But true, manual text-transformations are probablem way more KISS :D

**Reply by wojciech-kulik** [reply]
Interesting, thank you for the code example :)! I will see which way to go. Thanks for the help! 🍻

**Reply by L3MON4D3** [reply]
Happy I could help :)
